samples/csharp/web/avatar/Controllers/AvatarController.cs

```diff
@@ -0,0 +1,758 @@
+        {
+            { "role", "user" },
+            { "content", userQuery }
+        };
+
+            messages.Add(chatMessage);
+
+            // Stop previous speaking if there is any
+            if (isSpeaking)
+            {
+                await StopSpeakingInternal(clientId);
+            }
+
+            // For 'on your data' scenario, chat API currently has long (4s+) latency
+            // We return some quick reply here before the chat API returns to mitigate.
+            if (dataSources.Count > 0 && _clientSettings.EnableQuickReply)
+            {
+                await SpeakWithQueue(ClientSettings.QuickReplies[new Random().Next(ClientSettings.QuickReplies.Count)], 2000, clientId);
+            }
+
+            var url = $"{_clientSettings.AzureOpenAIEndpoint}/openai/deployments/{azureOpenaiDeploymentName}/chat/completions?api-version=2023-06-01-preview";
+            var body = JsonConvert.SerializeObject(new
+            {
+                messages,
+                stream = true
+            });
+
+            if (dataSources.Count > 0)
+            {
+                url = $"{_clientSettings.AzureOpenAIEndpoint}/openai/deployments/{azureOpenaiDeploymentName}/extensions/chat/completions?api-version=2023-06-01-preview";
+                body = JsonConvert.SerializeObject(new
+                {
+                    dataSources,
+                    messages,
+                    stream = true
+                });
+            }
+
+            var requestMessage = new HttpRequestMessage(HttpMethod.Post, url);
+            requestMessage.Headers.Add("api-key", _clientSettings.AzureOpenAIAPIKey);
+            requestMessage.Content = new StringContent(body, Encoding.UTF8, "application/json");
+
+            var response = await httpClient.SendAsync(requestMessage, HttpCompletionOption.ResponseHeadersRead);
+
+            if (!response.IsSuccessStatusCode)
+            {
+                throw new Exception($"Chat API response status: {response.StatusCode} {response.ReasonPhrase}");
+            }
+
+            var responseContent = await response.Content.ReadAsStringAsync();
+
+            // Process the responseContent as needed
+            var assistantReply = new StringBuilder();
+            var toolContent = new StringBuilder();
+            var spokenSentence = new StringBuilder();
+
+            var lines = responseContent.Split(separator, StringSplitOptions.None);
+
+            foreach (var line in lines)
+            {
+                if (line.StartsWith("data:") && !line.Contains("[DONE]"))
+                {
+                    try
+                    {
+                        var responseJson = JsonConvert.DeserializeObject<dynamic>(line[5..].Trim());
+                        var choices = responseJson!.choices;
+
+                        foreach (var choice in choices)
+                        {
+                            var delta = choice.delta;
```

Comment:
The json parsing logic is different when "dataSources.Count > 0" (on-your-data is enabled)
Need update code to cover this.
I'll give you the search key for testing this part.